dozer-api/src/cache_builder/mod.rs

```diff
@@ -1,295 +1,299 @@
+                    tokio::time::sleep(READ_LOG_RETRY_INTERVAL).await;
+                }
             }
         }
     }
 
-    Ok(())
+    async fn reconnect(&mut self) -> Result<(), ApiInitError> {
+        // Get endpoint meta.
+        let endpoint_meta =
+            EndpointMeta::new(&mut self.client, self.log_reader_options.endpoint.clone()).await?;
+        // Compare cache and log id.
+        self.state.update(
+            endpoint_meta.clone(),
+            &*self.cache_manager,
+            self.labels.clone(),
+            self.cache_write_options,
+        )?;
+        // Create log reader.
+        self.log_reader = create_log_reader(
+            &mut self.client,
+            endpoint_meta.schema,
+            self.log_reader_options.clone(),
+            self.state.next_log_position(),
+        )
+        .await?;
+        Ok(())
+    }
 }
 
-fn send_upsert_result(
-    endpoint_name: &str,
-    operations_sender: &Sender<GrpcOperation>,
-    upsert_result: UpsertResult,
-    schema: &Schema,
-    old: Option<Record>,
-    new: Record,
-) {
-    match upsert_result {
-        UpsertResult::Inserted { meta } => {
-            let op = types_helper::map_insert_operation(
-                endpoint_name.to_string(),
-                CacheRecord::new(meta.id, meta.version, new),
-            );
-            send_and_log_error(operations_sender, op);
-        }
-        UpsertResult::Updated { old_meta, new_meta } => {
-            // If `old` is `None`, it means `Updated` comes from `Insert` operation.
-            // In this case, we can't get the full old record, but the fields in the primary index must be the same with the new record.
-            // So we create the old record with only the fields in the primary index, cloned from `new`.
-            let old = old.unwrap_or_else(|| {
-                let mut record = Record::new(vec![Field::Null; new.values.len()]);
-                for index in schema.primary_index.iter() {
-                    record.values[*index] = new.values[*index].clone();
-                }
-                record
-            });
-            let op = types_helper::map_update_operation(
-                endpoint_name.to_string(),
-                CacheRecord::new(old_meta.id, old_meta.version, old),
-                CacheRecord::new(new_meta.id, new_meta.version, new),
-            );
-            send_and_log_error(operations_sender, op);
+fn open_or_create_cache(
+    cache_manager: &dyn RwCacheManager,
+    endpoint_meta: EndpointMeta,
+    labels: Labels,
+    write_options: CacheWriteOptions,
+) -> Result<Box<dyn RwCache>, ApiInitError> {
+    let (alias, cache_labels) = endpoint_meta.cache_alias_and_labels(labels.clone());
+    match cache_manager
+        .open_rw_cache(alias, cache_labels, write_options)
+        .map_err(ApiInitError::OpenOrCreateCache)?
+    {
+        Some(cache) => {
+            debug_assert!(cache.get_schema().0 == endpoint_meta.schema.schema);
```

Comment:
I think these should be regular `assert!`s or `Err` results, as if the schema is inconsistent, we'll be serving inconsistent data. Opening caches isn't on the hot path, so perf overhead should not have an impact on overall perf.